host-bin/startxiwi

```diff
@@ -0,0 +1,93 @@
+# found in the LICENSE file.
+
+set -e
+
+APPLICATION="${0##*/}"
+OPTS_ENTER=''
+OPTS_XIWI=''
+USAGE="$APPLICATION [options] CHROOT_APP [options]
+
+Wraps enter-chroot to launch a xiwi-app in a new windowed or tabbed session.
+You can launch any individual chroot graphical application in crouton windows.
+
+By default, it will log into the primary user on the first chroot found.
+
+Enter '-h' for 'enter-chroot' options and enter '-H' for 'xiwi-app' options.
+"
+USAGE_ENTER="enter-chroot [options]
+
+  -b          Fork and run the specified command silently in the background.
+  -c CHROOTS  Directory the chroots are in. Default: /mnt/stateful_partition/crouton/chroots
+  -l          Make the command part of a login. Parameters are passed directly
+              to the chroot command, and a call to su is appended.
+  -k KEYFILE  Override the auto-detected encryption key location.
+  -n NAME     Name of the chroot to enter. Default: first one found in CHROOTS
+  -t TARGET   Only enter the chroot if it contains the specified TARGET.
+  -u USERNAME Username (or UID) to log into. Default: 1000 (the primary user)
+  -X XMETHOD  (Override the auto-detected XMETHOD for this session.) - deprecated, 'xiwi' is enforced.
+  -x          Does not log in, but directly executes the command instead.
+              Note that the environment will be empty (sans TERM).
+              Specify -x a second time to run the /prepare.sh script.
+"
+USAGE_XIWI="xiwi [options] CHROOT_APP [options]
+
+  -f          Prevent xiwi from quitting automatically.[1]
+  -F          Will launch the app full-screen.[2]
+  -T          Will launch the app in a tab.
+
+By default, the CHROOT_APP is launched in a window.
+
+  [1] xiwi will normally close when the application returns. Some gui applications
+      fork before or during normal operation, which can confuse xiwi and cause it to
+      quit prematurely. If your application does not have a parameter that prevents
+      it from forking, and crouton is unable to automatically detect the fork, you can
+      use -f to prevent xiwi from quitting automatically.
+      xiwi will quit if you close the Chromium OS window when nothing is displayed.
+  
+  [2] A default window manager will full-screen all windows, unless CHROOT_APP begins
+      with 'start' or is 'xinit'. You can use -F to force the CHROOT_APP full-screen.
+
+  You can cycle through multiple windows inside the
+  application via Ctrl-Alt-Tab/Ctrl-Alt-Shift-Tab, or close them via
+  Ctrl-Alt-Shift-Escape.  If CHROOT_APP begins with 'start' but you still want to
+  use the default window manager, specify the full path of the application.
+"
+
+# Exits the script with exit code $1, spitting out message $@ to stderr
+error() {
+    local ecode="$1"
+    shift
+    echo "$*" 1>&2
+    exit "$ecode"
+}
+
+options=':bc:hlk:n:t:u:xfFHT'
+while getopts $options OPT; do
+  case ${OPT} in
+    b) OPTS_ENTER="$OPTS_ENTER -b";;
+    c) OPTS_ENTER="$OPTS_ENTER -c ${OPTARG}";;
```

Comment:
Also, OPTARG/OPTS_ENTER is very tricky here, as you need to be able to handle paths with spaces and whatnot.  One trick is to do what enter-chroot does, and prepare a string with each parameter '-escaped.  That means surrounding ${OPTARG} with single quotes, and passing the contents of OPTARG through sed to replace instances of single-quote with the line `'\''` (see enter-chroot).  Finally, when you go to actually call enter-chroot, you need to actually wrap it in an eval, so that the resulting OPTS_ENTER gets expanded and split.  HOWEVER, you need to make sure nothing else gets doubly-expanded (otherwise you'd have to manually escape that as well), so you need to escape out any other variable references in the eval line.

Basically, it's messy, and another reason why you want to combine all the OPTS_ENTER/OPTARG cases so you only have to do messy stuff once here, and once down below.